packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol

```diff
@@ -0,0 +1,285 @@
+        uint256 windowEnd,
+        address rewardToken,
+        bytes32 merkleRoot
+    ) external nonReentrant() onlyOwner {
+        uint256 indexToSeed = lastSeededIndex;
+        lastSeededIndex = indexToSeed.add(1);
+
+        _seedWindow(indexToSeed, totalRewardsDistributed, windowStart, windowEnd, rewardToken, merkleRoot);
+    }
+
+    // Flip window lock to opposite of current value. When `window.locked` is true,
+    // claims are blocked for that window.
+    function toggleWindowLock(uint256 windowIndex) external nonReentrant() onlyOwner {
+        merkleWindows[windowIndex].locked = !merkleWindows[windowIndex].locked;
+    }
+
+    // Overwrite merkle root for specified window index. Owner must deposit `totalRewardsDistributed`
+    // into contract.
+    // TODO: Should we require that `windowIndex <= lastSeededIndex` or should this function just be
+    // a generalized `setWindowMerkleRoot` whereby the owner can set a merkle root for any index?
+    function resetWindowMerkleRoot(
+        uint256 windowIndex,
+        uint256 totalRewardsDistributed,
+        uint256 windowStart,
+        uint256 windowEnd,
+        address rewardToken,
+        bytes32 merkleRoot
+    ) external nonReentrant() onlyOwner {
+        _seedWindow(windowIndex, totalRewardsDistributed, windowStart, windowEnd, rewardToken, merkleRoot);
+    }
+
+    // Emergency methods to transfer rewards into and out of the contract
+    // incase the contract was configured improperly.
+    function withdrawRewards(address rewardCurrency, uint256 amount) external nonReentrant() onlyOwner {
+        IERC20(rewardCurrency).safeTransfer(msg.sender, amount);
+    }
+
+    function depositRewards(address rewardCurrency, uint256 amount) external nonReentrant() onlyOwner {
+        IERC20(rewardCurrency).safeTransferFrom(msg.sender, address(this), amount);
+    }
+
+    /****************************
+     *
+     * Public functions
+     *
+     ****************************/
+
+    // TODO: This method could get pretty gas intensive; is there a way we can reduce the amount
+    // of external `transfer` calls if we precompute all the rewards for each account?
+    // It's a bit tricky because there are multiple reward currencies possible.
+    function claimWindows(Claim[] memory claims) public nonReentrant() {
+        for (uint256 i = 0; i < claims.length; i++) {
+            _claimWindow(claims[i]);
+        }
+    }
+
+    // Claim `amount` of reward tokens for `account`. If `amount` and `account` do not exactly match the values stored
+    // in the merkle proof for this `windowIndex` this method will revert.
+    function claimWindow(Claim memory claim) public nonReentrant() {
+        _claimWindow(claim);
+    }
+
+    // Checks {account, amount} against Merkle root at given window index.
+    function verifyClaim(Claim memory claim) public view returns (bool valid) {
+        bytes32 leaf = keccak256(abi.encodePacked(claim.account, claim.amount));
```

Comment:
i want to be sure here, but you cannot call claim with an amount other than what was original encoded in the merkle proof? Even though it might make sense to claim less than what you are owed, that is not possible with this?

Reply:
>i want to be sure here, but you cannot call claim with an amount other than what was original encoded in the merkle proof

Correct

> Even though it might make sense to claim less than what you are owed, that is not possible with this?
This could be possible if we allowed the user to pass in an additional parameter `amountToClaim` to `claimWindow`. They would still have to include the `claim.amount` to verify the `merkleProof`, but they could also specify that they only want to receive `amountToClaim`. Then we can increment `amountClaimedFromWindow` accordingly. Do you think this is a worthwhile feature? @mrice32 @chrismaree 

Personally I'm not sure that partial claims are a useful feature. I'm not even convinced vesting is worth the added complexity. Some of the vesting logic can be repurposed to handle partial claims however.

Reply:
No im not suggesting partial claims, unecessary complexity. Was just trying to work through the logic to find edge cases

Reply:
The more I'm thinking about this the more i'm thinking we should allow partial claims. It does add more complexity and adds a couple more test cases, but shouldn't be much more smart contract code. PieDAO's vesting-enabled distributor does something like this where the user specifies both a `_amount` and a `_windowAmount` (equivalent to our claim.amount): https://github.com/pie-dao/vested-token-migration-app/blob/master/contracts/VestedTokenMigration.sol#L71

Reply:
in what scenario would you want partial claims?

Reply:
Allows more flexibility to users? Might end up being an unused feature with added complexity.